### PR TITLE
feat(nimbus): Fetch A-S megazords from Taskcluster

### PR DESCRIPTION
**Pull request by brennie**
Because

- we don't want to manually build application-services megazords
  ourselves for each container; and
- application-services publishes megazord artifacts on Taskcluster

This commit

- adds a new as-builder docker container to download cirrus and nimbus
  megazords with the appropriate binaries for the platform as Python
  packages;
- updates the cirrus and experimenter containers to copy the megazords
  from the as-builder container;
- updates cirrus and experimenter to refer to the new megazord packages;
- updates cirrus to be based on Debian Bullseye so it can use the
  generated binaries; and
- updates the Dockerfile-rust-image container to be built on top of
  as-builder in CI.

### CODE REVIEW

File: application-services/fetch-application-services.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  local LIB_DIR="/tmp/${PROJECT}/${TARGET}"
+
+  if [[ ! -d "${LIB_DIR}" ]]; then
+    echo >&2 "Could not find libraries for target ${TARGET} at ${LIB_DIR}"
+    exit 1
+  fi
+
+  mkdir -p "${PROJECT}/lib"
+  cp /tmp/"${PROJECT}"/*.py "${PROJECT}"
+  cp /tmp/"${PROJECT}"/"${TARGET}"/* "${PROJECT}/lib"
```
Comment (jhugman) [inline application-services/fetch-application-services.sh:35]: I think the `.so` file needs to be in the same directory as the python files.

Also, consider moving the additional python files (in `src` and `lib`, if present) at this point.

Suggest 

```suggestion
local SO_FILE=$(/tmp/$PROJECT/scripts/detect-target.sh $PROJECT)
cp "/tmp/$PROJECT/$TARGET/$SO_FILE" "$PROJECT"
```

File: cirrus/server/Dockerfile [hunk 2]
```diff
@@ -51,15 +26,15 @@ COPY poetry.lock /cirrus/
...
 WORKDIR /cirrus
 
-# Copy specific Rust components
-COPY --from=rust-builder /application-services/target/release/libcirrus.so /application-services/libcirrus.so
-COPY --from=rust-builder /application-services/cirrus.py /application-services/cirrus_sdk.py
-COPY --from=rust-builder /application-services/fml.py /application-services/fml_sdk.py
+# Copy cirrus megazord
+COPY --from=as-builder /application-services/cirrus/lib/libcirrus.so /application-services/libcirrus.so
+COPY --from=as-builder /application-services/cirrus/cirrus.py /application-services/cirrus_sdk.py
+COPY --from=as-builder /application-services/cirrus/fml.py /application-services/fml_sdk.py
```
Comment (jhugman) [inline cirrus/server/Dockerfile:37]: I think splitting the logic for preparing the megazord between here and `fetch-application-services.sh` can be avoided, so we just have

```suggestion
COPY --from=as-builder /application-services/cirrus /application-services
```

or 

```suggestion
COPY --from=as-builder /application-services/cirrus /application-services/megazord
```

File: experimenter/Dockerfile [hunk 3]
```diff
@@ -111,14 +92,13 @@ COPY --from=node-builder /experimenter/experimenter/theme/static_src/node_module
...
-# The uniffi.toml file setting cdylib_name is set to "cirrus", which means the generated python is expecting
-# the library to be called libcirrus.so.
-COPY --from=rust-builder /application-services/target/release/libnimbus_experimenter.so /application-services/megazord/libcirrus.so
+# Copy nimbus-experimenter megazord
+COPY --from=as-builder /application-services/nimbus-experimenter/fml.py /application-services/megazord/rust_fml.py
+COPY --from=as-builder /application-services/nimbus-experimenter/fml.py /application-services/megazord/rust_fml.py
+COPY --from=as-builder /application-services/nimbus-experimenter/nimbus.py /application-services/megazord/rust_sdk.py
+COPY --from=as-builder /application-services/nimbus-experimenter/errorsupport.py /application-services/megazord/rust_errorsupport.py
+COPY --from=as-builder /application-services/nimbus-experimenter/remote_settings.py /application-services/megazord/rust_remote_settings.py
+COPY --from=as-builder /application-services/nimbus-experimenter/lib/libnimbus_experimenter.so /application-services/megazord/libnimbus_experimenter.so
```
Comment (jhugman) [inline experimenter/Dockerfile:101]: Again, let's do all the preparation of the megazord in the `fetch-application-services.sh` script, so we're just left with

```suggestion
COPY --from=as-builder /application-services/nimbus-experimenter /application-services/megazord
```

File: cirrus/server/Dockerfile [hunk 4]
```diff
@@ -1,30 +1,5 @@
...
-
-# Checkout application services at a locked commit
-RUN git clone https://github.com/mozilla/application-services.git .
-RUN git fetch && git checkout da9301f829de67149d50dce5b709aed42f1bfe1b
-
-RUN git submodule init
-RUN git submodule update --recursive
-
-RUN cargo build --manifest-path megazords/cirrus/Cargo.toml --release
-RUN cargo uniffi-bindgen generate --library target/release/libcirrus.so --language python --out-dir .
```
Comment (jhugman) [inline cirrus/server/Dockerfile:24]: 😻

**Review by jhugman — COMMENTED** [review]
This is so good to see at last. Thank you for pushing it over the line.

I can't comment on the `Docker` bits, but I have put some suggestions for preparing the megazord from the downloaded artifacts.

Also: is there still more to do? i.e. the `Dockerfile-rust-image`

**Comment by brennie** [comment]
> This is so good to see at last. Thank you for pushing it over the line.
> 
> I can't comment on the `Docker` bits, but I have put some suggestions for preparing the megazord from the downloaded artifacts.
> 
> Also: is there still more to do? i.e. the `Dockerfile-rust-image`

I think the changes to Dockerfile-rust-image and its CI steps are correct, but I'll need @jrbenny35 to weight in there.

I've also updated the megazords to be Python packages (by adding a blank `__init__.py` in the megazord directory) and then updated Cirrus and Experimenter to reference the packages instead of the free-floating python files.

**Review by jaredlockhart — APPROVED** [review]
Lol @brennie I was gonna do all this but you went and did it anyways, thank you 🙏 ❤️  This is exactly what I was thinking, great job all looks great.  Aaaaand fantastic we have the Cirrus integration test now to be able to do refactors like this safely.  All looks exactly as I expected, great job 🎉

File: .circleci/config.yml [hunk 5]
```diff
@@ -591,34 +591,14 @@ jobs:
...
           name: Build rust test image
           command: |
             set +e
             docker kill nimbus-rust-image
             docker rm nimbus-rust-image
-            AS_VERSION=$(git -c 'versionsort.suffix=-' \
-              ls-remote --exit-code --refs --sort='version:refname' --tags https://github.com/mozilla/application-services.git \
-              '*.*.*' | tail --lines=1 | cut --delimiter='/' --fields=3)
+            make build_as
             docker buildx build -t ${DOCKERHUB_REPO}:nimbus-rust-image -f experimenter/tests/integration/nimbus/utils/Dockerfile-rust-image --build-arg as_version=$AS_VERSION --progress=plain .
```
[the remark this answers is not in the document]
Comment (brennie) [reply .circleci/config.yml:601]: Good call!

**Pull request merged**